test/behaviour/resolution/correctness/SoundnessVerifier.java

```diff
@@ -0,0 +1,112 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.test.behaviour.resolution.correctness;
+
+import com.vaticle.typedb.core.TypeDB;
+import com.vaticle.typedb.core.TypeDB.Transaction;
+import com.vaticle.typedb.core.common.parameters.Arguments;
+import com.vaticle.typedb.core.common.parameters.Options;
+import com.vaticle.typedb.core.concept.Concept;
+import com.vaticle.typedb.core.concept.answer.ConceptMap;
+import com.vaticle.typedb.core.reasoner.resolution.answer.Explanation;
+import com.vaticle.typedb.core.test.behaviour.resolution.correctness.CorrectnessVerifier.SoundnessException;
+import com.vaticle.typedb.core.test.behaviour.resolution.correctness.Materialiser.Materialisation;
+import com.vaticle.typedb.core.traversal.common.Identifier.Variable.Retrievable;
+import com.vaticle.typeql.lang.query.TypeQLMatch;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+class SoundnessVerifier {
+
+    private final Materialiser materialiser;
+    private final TypeDB.Session session;
+    private final Map<Concept, Concept> inferredConceptMapping;
+    private final Set<Explanation> verified;
+
+    private SoundnessVerifier(Materialiser materialiser, TypeDB.Session session) {
+        this.materialiser = materialiser;
+        this.session = session;
+        this.inferredConceptMapping = new HashMap<>();
+        this.verified = new HashSet<>();
+    }
+
+    static SoundnessVerifier create(Materialiser materialiser, TypeDB.Session session) {
+        return new SoundnessVerifier(materialiser, session);
+    }
+
+    void verifyQuery(TypeQLMatch inferenceQuery) {
+        try (Transaction tx = session.transaction(Arguments.Transaction.Type.READ,
+                                                  new Options.Transaction().infer(true).explain(true))) {
+            tx.query().match(inferenceQuery).forEachRemaining(ans -> verifyAnswer(ans, tx));
+        }
+    }
+
+    private void verifyAnswer(ConceptMap answer, Transaction tx) {
+        answer.explainables().iterator().forEachRemaining(explainable -> {
+            tx.query().explain(explainable.id()).forEachRemaining(explanation -> {
+                // This check is valid given that there is no mechanism for recursion termination given by the UX of
+                // explanations, so we do it ourselves
+                if (verified.contains(explanation)) return;
+                else verified.add(explanation);
+                verifyAnswer(explanation.conditionAnswer(), tx);
+                verifyExplanation(explanation);
```

Comment:
hmm yeah I suppose it reads ok here, but less well in the `Completeness` checker i think!